### DISCUSSION TITLE
(How) can I customize the various pages that all use the blog layout? Can I define "sublayouts" or pass parameters?

### DISCUSSION THREAD

**Discussion by cellio** [datenstrom/community#713] [Ask a question] [answered]
The blog layout, of course, uses the `system/layouts/blog-start.html` layout file.  Two other layouts also use this: specific tags and specific months in the archive.  That is, if I click on a tag, or on a month in the archive, what I get is a filtered page (just that tag or just that month) using the blog-start layout.  There isn't a markdown file that I can edit for tags and months; the code takes care of that.

I would like to add some customizations to these other pages.  For tag pages, I'd like to be able to add a description of the tag to the top of the page (possibly linking to related tags).  For months, I'd like to add "previous month" and "next month" links to the bottom of the page (which I asked about in https://github.com/datenstrom/yellow/discussions/700).  What I want to do, effectively, is to either "subclass" blog-start (tagpages and archivepages) or modify blog-start to adjust content based on the "type" of blog-start page it is.

I think (but I'm not sure, because I don't know PHP) that I could find the part of the code that generates tag pages and archive pages, find where it specifies "blog-start", and change those to new layouts that I would then create.  But I think I would have to clone blog-start to make the new ones, which means that if I ever want to change blog-start in the future, I'd have to do it in three places.  Is there actually a way to share content among layouts (some sort of inheritance or include)?

Alternatively, it seems like it ought to be possible to modify the code for these pages to still use blog-start but somehow pass the type of page along, and the blog-start layout could use conditionals to make these small additions.  I've seen conditionals in the layouts, so I know that's possible, but I don't know how to pass the argument.  I also don't know where I would add those tag descriptions in the content directory.

Am I on the right track?  Is either of these approaches possible, or is there a better way to do what I'm trying to accomplish?

**Comment by GiovanniSalmeri** [comment, marked as answer]
>  it seems like it ought to be possible to modify the code for these pages to still use blogpages but somehow pass the type of page along, and the blogpages layout could use conditionals to make these small additions. I've seen conditionals in the layouts, so I know that's possible, but I don't know how to pass the argument.

I think you could add in the page a setting (in the section between `---`), for example:

```
PageType: verybeautiful
```

and then in the layout you can put a conditional:

```php
<php if ($this->yellow->page->get("pageType")=="verybeautiful"): ?>
....
<?php endif; ?>
```

**Reply by markseuffert** [reply]
Yes that's true. As a developer all you need is a basic knowledge of HTML, CSS and some help of the API. You can access information that are stored in the file system and communicated via HTTP. A good starting point is the source code of the extension you're using and the [API for developers](https://datenstrom.se/yellow/help/api-for-developers). Here are some examples:

* `$this->yellow->page->get("pageType")`: Return page setting, in this case a custom type
* `$this->yellow->page->getPage("footer")`: Return shared page, in this case a page footer
* `$this->yellow->page->getRequest("tag")`: Return page request argument, in this case the current tag

To show a description of tags change the following in your layout file `system/layouts/blog-start.html`:

```
...
<?php if ($this->yellow->page->get("blogWithFilter")): ?>
<h1><?php echo $this->yellow->page->getHtml("titleContent") ?></h1>
<?php echo $this->yellow->page->getPage($this->yellow->page->getRequest("tag"))->getContent() ?>
<?php endif ?>
...
```

To create a description for the tag `example` create the file `content/shared/example.md`. Here's an example:

```
---
Title: Example
Status: shared
---
Build projects around motivated individuals. Give them the environment and
support they need, and trust them to get the job done. This text is taken from
the Agile Manifesto and I believe it's still valid for [tech](/blog/tag:tech/)
and [online communities](/blog/tag:online-communities/).
```
[Screenshot for the curious](https://user-images.githubusercontent.com/2429165/155164157-f99f8ba5-1066-44dd-9a37-8e665364ab26.png). Tested with Datenstrom Yellow 0.8.19.

**Reply by cellio** [reply]
Thank you!  I appreciate the detailed example.
